### PR TITLE
UI for multilevel cache implemented to implement Dinero

### PR DESCRIPTION
**Pull request by ismailfaruk**
UI features added to make using Dinero more user friendly. Scroll button added. Container/frame resized for better UX/UI. Internal components resized.

### CODE REVIEW

**Comment by lupino3** [comment]
Hi Ismail,
thanks for your contribution! I like the new UI and would like to merge your code into the master branch.

However, I have a few suggestions to improve your code, I hope you'll be able to implement them before merging. Let me know if you have any questions about any comment or if you cannot / do not want to implement some of them and why, and we can discuss them :)

Thanks again for your contribution, looking forward to the new UI!
Andrea

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 1]
```diff
@@ -137,10 +144,110 @@ public void keyReleased(KeyEvent e) {
...
+		//combo box action of cache type - passing combo option to argument for cache panel
+		cacheType.addActionListener(new ActionListener() {
+			public void actionPerformed(ActionEvent e){
+				argType = (String.valueOf(cacheType.getSelectedItem())).charAt(0);
+			}
+		});
+		cacheType.setSelectedIndex(2);
+
+		//button for creating cache defined
+		create = new JButton("Create Cache");
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: You will need to add `create` to the container `cp`, otherwise it is not visible.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 2]
```diff
@@ -137,10 +144,110 @@ public void keyReleased(KeyEvent e) {
       }
     });
 
-    JButton browse = new JButton("Browse...");
+		String[] level = {"1", "2", "3", "4", "5"};
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: Please use indentation consistent with the rest of the file (2 spaces to indent). Otherwise it becomes hard to follow the flow of the file if it uses mixed indentation.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 3]
```diff
@@ -46,8 +46,15 @@
   // one instance of DineroFrame will be created in EduMIPS64
   private static final Logger logger = Logger.getLogger(DineroFrontend.class.getName());
   private static JTextField path, params;
-  private static JButton execute;
+  private static JButton execute, configure, create;	//configure and create buttons added to create panels and cache parameters
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: It would be nice to not need a `configure` action at all, but rather re-create dynamically the command line every time the user interacts with any of the UI elements.

Note: this is a larger and optional change, feel free to address it in a subsequent PR.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 4]
```diff
@@ -46,8 +46,15 @@
...
   private static JTextArea result;
+  private static JLabel pathLabel, paramsLabel, cacheLeveLabel, cacheTypeLabel;	//labels for cache level and cache type added
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: `pathLabel` and `paramsLabel` are not used, please remove them.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 5]
```diff
@@ -46,8 +46,15 @@
...
+  private static int argLevel;	//static declaration of array Level argument
+	private static char argType;	//static declaration of array type argument
+	private static Box cachePanel;	//static container globally declared for flexibility
+	private static JScrollPane scrollPane;	//static delclaration of scrollbar
+  public static JPanel panelL1, panelL2, panelL3, panelL4, panelL5;	//static panels to add configuration components
+  private static JComboBox cacheLevel, cacheType;		//static delcaration of cache combo boxes
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: `JComboBox` is a generic class (https://docs.oracle.com/javase/8/docs/api/javax/swing/JComboBox.html). Please use the generic version (in this case, `JComboBox<String>`).

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 6]
```diff
@@ -243,15 +350,165 @@ public void keyReleased(KeyEvent e) {
...
+		//Resized the frame container to have better UX with the new panels
+		setSize(900, 800);
   }
 }
+
+/** Panel with all the necessary controls to modify the options of a Cache.
+ */
+class DineroSingleCachePanel extends JPanel {
+	private DineroCacheOptions dco;
+	private JComboBox size, sizeUnit, bsize, bsizeUnit, level, type;
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: Same here, please use `JComboBox<String>`.

Also, looks like `lwvel` and `type` are not used, please remove them.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 7]
```diff
@@ -243,15 +350,165 @@ public void keyReleased(KeyEvent e) {
...
+	private JTextField assoc;
+	private JCheckBox ccc;
+	private JLabel cacheSizeLabel, cacheSizeUnitLabel, blockSizeLabel, bsizeUnitLabel, assocLabel, cccLabel;
+
+	public DineroSingleCachePanel(char type, int level) {
+		dco = new DineroCacheOptions(type, level);
+
+		String[] sizes = {"1", "2", "4", "8", "16", "32", "64", "128", "256", "512"};
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: Consider making those arrays `static final` members of the class.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 8]
```diff
@@ -243,15 +350,165 @@ public void keyReleased(KeyEvent e) {
...
+	
+	private char type;
+	private int level;
+	
+	public DineroCacheOptions(char type, int level) {
+		this.type = type;
+		this.level = level;
+	}
+	
+	public String toString() {
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: Consider using `StringBuilder` here instead of simple string concatenation, which allocates new strings every time a concatenation is done.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 9]
```diff
@@ -243,15 +350,165 @@ public void keyReleased(KeyEvent e) {
...
+	//Passes Dinero command parameters with proper syntax
+	public String toString(){
+		ccc.addItemListener(new ItemListener(){
+			public void itemStateChanged(ItemEvent e) {
+				Boolean cccFlag =  ccc.isSelected();
+				dco.ccc = cccFlag;
+			}
+		});
+
+		//Implements try and catch method to only allow numerical value only
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: This try/catch could be moved inside `DineroCacheOptions.toString()`, since this functionality of checking data before conversion logically belongs to `DineroCacheOptions`.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 10]
```diff
@@ -243,15 +350,165 @@ public void keyReleased(KeyEvent e) {
...
+
+		//Default dimension set for panel
+		setPreferredSize(new Dimension(850, 80));
+		setMaximumSize(new Dimension(850, 90));
+		setMinimumSize(new Dimension(850, 50));
+	}
+
+	//Passes Dinero command parameters with proper syntax
+	public String toString(){
+		ccc.addItemListener(new ItemListener(){
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: Why is this listener set in the `toString` method? I think it belongs to the constructor.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 5 above)
Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: The changes have been made. Thank you.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 7 above)
Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: The changes have been made. Thank you.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 6 above)
Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: The changes have been made. Thank you.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 1 above)
Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: Added a few more labels along with `create`. It will be visible along with the new PR.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 3 above)
Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: I agree. I will address this in another PR and update here later on.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 5 above)
Comment (lupino3) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: I don't see the change.. did you push your changes?

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 1 above)
Comment (lupino3) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: What "new PR"? Do you intend to abandon this PR and submit a new one, or to push more commits to this PR?

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 11]
```diff
@@ -500,15 +515,29 @@ public DineroCacheOptions(char type, int level) {
...
-		if(assoc > 0)
-			cmdline += prefix + "assoc" + " " + assoc + " ";
-		if(ccc)
-			cmdline += prefix + "ccc" + " ";
+		// creates empty builder, capacity 16
+		StringBuilder cacheConfig = new StringBuilder();
+		// adds 9 character string at beginning
+		try {
+			String prefix = "-l" + level + "-" + type;
+			cacheConfig.append("prefix");
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: This will just append to the stringbuilder the string `"prefix"`, not the variable `prefix`.

Also, `String.format` is probably a better way to create the options instead of concatenation.

So for example the prefix can be created and appended this way:

```
String prefix = String.Format("-l%d-%s", level, type);

cacheConfig.append(prefix);
```

To this, you can then create the other parts of the command line and append them (size, bsize, etc). No need to add them to cmdline any more.

For example, for size:

```
cacheConfig.append(String.Format("size %s", size));
```

Does this make sense?

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java [hunk 12]
```diff
@@ -500,15 +515,29 @@ public DineroCacheOptions(char type, int level) {
...
+				cmdline += prefix + "assoc" + " " + assoc + " ";
+			if(ccc)
+				cmdline += prefix + "ccc" + " ";
+			
+			cacheConfig.append("cmdline");
+			return cmdline;
+		}
 
-		return cmdline;
+		catch (Exception e) {
```
Comment (lupino3) [inline src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: Whenever you catch exceptions, at least log them so we know what went well. Ideally, capture specific sub-classes of exception and handle them in different ways, if it makes sense.

Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: Yes, I agree. I plan to update it similar to the ones implemented before. Thank you for pointing it out.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 11 above)
Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: I have done the string builder part much, but I will update once I finish it. I recall it was a previous change that you requested. But for now, I think I understand what you have mentioned here.

**Comment by lupino3** [comment]
Hey @ismailfaruk, any updates on finishing up this PR?

**Comment by ismailfaruk** [comment]
> 
> 
> Hey @ismailfaruk, any updates on finishing up this PR? Let me know if you need anything from me to help you get it done! Thanks!

**Comment by ismailfaruk** [comment]
Hello

Sorry, I was away from working on GitHub for a while. I will get back on
this asap. And will definitely ask for help if needed. Thanks.

**Comment by lupino3** [comment]
Hi Ismael,
sure, thank you very much. I'll reopen the pull request since we have some work to do before merging it. Thanks!

**Comment by lupino3** [comment]
Hello @ismailfaruk, I haven't heard from you in a while, so I am closing this PR. Feel free to reopen it if you have time to work on it and want to get it merged. I'll be happy to help you. Thanks!

**Comment by ismailfaruk** [comment]
Hi @lupino3 
Sorry I forgot to finish this. I will finish the things that you had mentioned above and once done, I will reopen the PR. Thanks for the message.

**Comment by lupino3** [comment]
Sounds good, looking forward to it. Thank you!

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 12 above)
Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java:539]: Tried using printStackTrace and toString to get the trace and print the error. Let me know if I should add more details to this.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 2 above)
Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: Tried to fix the indentation, will get back to this part again once the rest is done.

**Comment by ismailfaruk** [comment]
Hi, could you please reopen the PR? I cannot reopen the PR. I have not applied all changes but I just wanted to show them being made to get feedback.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 11 above)
Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: Would it be alright if I did not use StringBuilder at all? Because the use of the string is very limited and using an object may be less beneficial than just a String variable.

File: src/main/java/org/edumips64/ui/swing/DineroFrontend.java (hunk 9 above)
Comment (ismailfaruk) [reply src/main/java/org/edumips64/ui/swing/DineroFrontend.java]: I am thinking of moving it to the constructor rather, along with the ccc action listener.

**Comment by lupino3** [comment]
Closing given there hasn't been activity in the past few years. @ismailfaruk, please let me know when you're interested in resuming this work, I'll be happy to help you!
